source/includes/fundamentals/code-snippets/Monitoring.java

```diff
@@ -0,0 +1,169 @@
+package com.mycompany.app;
+
+import com.mongodb.ConnectionString;
+import com.mongodb.MongoClientSettings;
+import com.mongodb.ReadPreference;
+import com.mongodb.client.MongoClient;
+import com.mongodb.client.MongoClients;
+import com.mongodb.client.MongoCollection;
+import com.mongodb.client.MongoDatabase;
+import com.mongodb.event.*;
+import org.bson.Document;
+
+import java.util.concurrent.TimeUnit;
+
+public class Monitoring {
+
+    private static final String COLLECTION = "compound-ops";
+    private static final String DATABASE = "test";
+    private static final ConnectionString URI = new ConnectionString(System.getenv("DRIVER_URL"));
+
+    public static void main(String[] args) throws InterruptedException {
+        Monitoring examples = new Monitoring();
+        System.out.println("\n---Command Event---\n");
+        examples.monitorCommandEvent();
+        System.out.println("\n---Cluster Event---\n");
+        examples.monitorClusterEvent();
+        System.out.println("\n---Connection Pool Event---\n");
+        examples.monitorConnectionPoolEvent();
+    }
+
+    private void monitorCommandEvent() {
+        // start monitor-command-example
+        MongoClientSettings settings =
+                MongoClientSettings.builder()
+                        .applyConnectionString(URI)
+                        .addCommandListener(new CommandTimer())
+                        .build();
+        MongoClient mongoClient = MongoClients.create(settings);
+        // Run some commands to test the timer
+        MongoDatabase database = mongoClient.getDatabase(DATABASE);
+        MongoCollection<Document> collection = database.getCollection(COLLECTION);
+        collection.find().first();
+        collection.countDocuments();
+        mongoClient.close();
+        // end monitor-command-example
+    }
+
+    private void monitorClusterEvent() {
+        // start monitor-cluster-example
+        IsReadAndWrite clusterListener = new IsReadAndWrite();
+        MongoClientSettings settings =
+                MongoClientSettings.builder()
+                        .applyConnectionString(URI)
+                        .applyToClusterSettings(builder ->
+                                builder.addClusterListener(clusterListener))
+                        .build();
+        MongoClient mongoClient = MongoClients.create(settings);
+        MongoDatabase database = mongoClient.getDatabase(DATABASE);
+        MongoCollection<Document> collection = database.getCollection(COLLECTION);
+        // Run a command to trigger a ClusterDescriptionChangedEvent event
+        collection.find().first();
```

Comment:
Question: Your output says it's also able to write to the server. But my understanding of this event is that it only prints if an operation was performed. So the message about being able to read makes sense to me, but where does the write message come from?

Reply:
The write message comes from the `hasWritableServer()` method of the [`ClusterDescription`](https://mongodb.github.io/mongo-java-driver/4.2/apidocs/mongodb-driver-core/com/mongodb/connection/ClusterDescription.html#hasWritableServer()) object.

The [Cluster Description](https://mongodb.github.io/mongo-java-driver/4.2/apidocs/mongodb-driver-core/com/mongodb/connection/ClusterDescription.html#hasWritableServer()) has general information about the cluster the driver is connecting with. Is there something about this paragraph that is unclear or confusing?